File: 208_implement_trie_(prefix_tree).java
class Trie {
    private final int SIZE = 26;

    class Node {
        char c;
        Node[] childs;
        boolean isFound;

        Node(char c) {
            this.c = c;
            this.childs = new Node[SIZE];
            this.isFound = false;
        }
        Node() {
            this.childs = new Node[SIZE];
            this.isFound = false;
        }
    }

    Node root;

    private Node createTrieNode(char c) {
        Node node = new Node(c);
        return node;
    }

    /** Initialize your data structure here. */
    public Trie() {
        this.root = new Node();
    }

    /** Inserts a word into the trie. */
    public void insert(String word) {
        Node curr = this.root;
        Node prev = null;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            int idx = c - 'a';
            if (curr.childs[idx] == null) {
                curr.childs[idx] = createTrieNode(c);
            }
            curr = curr.childs[idx];
        }
        curr.isFound = true;
    }

    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        Node curr = this.root;
        int charIdx = 0;
        for (int i = 0; i < word.length() && curr != null; i++) {
            char c = word.charAt(i);
            curr = curr.childs[c - 'a'];
        }
        return curr != null && curr.isFound;
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        Node curr = this.root;
        int charIdx = 0;
        for (int i = 0; i < prefix.length() && curr != null; i++) {
            char c = prefix.charAt(i);
            curr = curr.childs[c - 'a'];
        }
        return curr != null;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: python/208_implement_trie(prefix_tree).py
class Node:
    def __init__(self):
        self.next = [None for _ in range(26)]
        self.string = None

class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = Node()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        curr = self.root
        for c in word:
            index = ord(c) - ord('a')
            if not curr.next[index]:
                curr.next[index] = Node()
            curr = curr.next[index]
        curr.string = word

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        curr = self.root
        for c in word:
            index = ord(c) - ord('a')
            if not curr.next[index]:
                return False
            curr = curr.next[index]
        return curr.string and curr.string == word

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        curr = self.root
        for c in prefix:
            index = ord(c) - ord('a')
            if not curr.next[index]:
                return False
            curr = curr.next[index]
        return True

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
